src/pdmat.jl

```diff
@@ -1,23 +1,49 @@
 """
 Full positive definite matrix together with a Cholesky factorization object.
 """
-struct PDMat{T<:Real,S<:AbstractMatrix} <: AbstractPDMat{T}
+struct PDMat{T<:Real,S<:AbstractMatrix{T}} <: AbstractPDMat{T}
     mat::S
     chol::Cholesky{T,S}
 
-    PDMat{T,S}(m::AbstractMatrix{T},c::Cholesky{T,S}) where {T,S} = new{T,S}(m,c)
+    function PDMat{T,S}(m::AbstractMatrix, c::Cholesky) where {T,S}
+        d = LinearAlgebra.checksquare(m)
+        if size(c, 1) != d
+            throw(DimensionMismatch("Dimensions of mat and chol are inconsistent."))
+        end
+        # in principle we might want to check that `c` is a Cholesky factorization of `m`,
+        # but that's slow
+        return new{T,S}(m,c)
+    end
 end
+PDMat{T,S}(pdm::PDMat) where {T,S} = PDMat{T,S}(pdm.mat, pdm.chol)
```

Comment:
Is it common to define these constructors with type parameters? Or even necessary? It seems it's not done for any of the other matrix types in PDMats. I assumed that defining the relevant `convert` methods and the optimization
```suggestion
PDMat(pdm::PDMat) = pdm
```
would be sufficient?

Reply:
To be clear, this is one of the elegances of Julia's type system. `Matrix{T}` is a concrete type, and you should be able to construct the concrete type of your choosing: you shouldn't be doomed to just use `Matrix(A)`, because that does not allow you to control the eltype. `Cholesky{T}` is not a concrete type, because Cholesky depends on two type parameters. A very cool aspect of Julia's type system is that you can specify the first concretely and allow the second to be computed from the inputs, *or* you can specify both concretely. Thus you want methods `Cholesky{T,S})(args...)`, `Cholesky{T}(args...)` and `Cholesky(args...)` (all 3 variants).

That's what this PR adds for `PDMat`. More of Julia's linear algebra ecosystem should follow suit, but you have to start somewhere.

Reply:
I see. I was not sure regarding the conventions for `AbstractMatrix` constructors and conversions and was wondering whether we could rely on some existing functionality in LinearAlgebra. Based on the linked PR, it seems the approach in this PR (defining constructors instead of `convert`) follows the convention in LinearAlgebra.